uint32 resolution
---
int32 return_code